VCDA-594: Add scripts to enable Jenkins multi-branch builds on commits

hodgesrm opened the pull request:
Implemented Jenkinsfile to drive multi-branch builds in Jenkins. Added
run_examples.sh script which automatically fills in tenant.yaml values
and executes samples. Added run_system_tests.sh which automatically edits
base_config.yaml values. Upgraded tenant-onboard.py example to select
pvdc and network pool values automatically.  Added basic README.md files
to explain run script behavior.

Reviewers: @rocknes, @sahithi, @sompa, @anusuyar

sahithi commented:
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[.Jenkinsfile, line 7 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOHveBDVAmsLrhm1n7:-LBOHveBDVAmsLrhm1n8:brl2cfc) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/.Jenkinsfile#L7)):*
> ```
> // Helper function to send email. 
> def sendEmail() {
>     if (currentBuild.result == currentBuild.previousBuild.result) {
> ```

Can you elaborate on the If condition here?

---

*[.Jenkinsfile, line 34 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOJRED6WL6sikzlHrW:-LBOISEL0LNNOelyhR5L:b-cuc61f) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/.Jenkinsfile#L34)):*
> ```
>                 // Remove and checkout current branch from git.
>                 deleteDir()
>                 checkout scm
> ```

What is checkout scm doing? git clone? If yes, where is this command or script located?

---

*[.Jenkinsfile, line 86 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOJj5wCFDIHmBc_ZKl:-LBOJj5wCFDIHmBc_ZKm:b-cmgpd3) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/.Jenkinsfile#L86)):*
> ```
>         stage('cleanup') {
>             steps {
>                 echo "We don't do any clean-up for now."
> ```

I think restoring the testbed environment should go here. 

---

*[.Jenkinsfile, line 96 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOK75a1SvF_RCC1cHN:-LBOK75a1SvF_RCC1cHO:b5k1isp) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/.Jenkinsfile#L96)):*
> ```
>         success { 
>             echo "Job succeeded; will post email on recovery from failure"
>             echo "This is where we can restore the vCD installation snapshot!"
> ```

Regardless of success or failure, restoration of testbed must happen.

---

*[examples/tenant.yaml, line 55 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOKxcgCtv8cIZu0UtW:-LBOKxcgCtv8cIZu0UtX:bk3tteb) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/examples/tenant.yaml#L55)):*
> ```YAML
>     description: "Test vApp"
>     catalog_name: "Test-Catalog"
>     file_name: "../system_tests/test_vapp_template.ova"
> ```

If test_vapp_template.ova is expected to be present, document it somewhere. It can save some debugging time in future for a new user.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/234)*
<!-- Sent from Reviewable.io -->

rocknes commented:
Reviewed 4 of 8 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[.Jenkinsfile, line 7 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOHveBDVAmsLrhm1n7:-LBWitRz3oFQtCRzmdmL:b-uqmgjo) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/.Jenkinsfile#L7)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Can you elaborate on the If condition here?
</blockquote></details>

I believe the email sending code isn't working as of now. Correct me if I am wrong @hodgesr

---

*[examples/tenant.yaml, line 55 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOKxcgCtv8cIZu0UtW:-LBWlVgP5qyT6lKvTiR4:b30vfrl) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/examples/tenant.yaml#L55)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

If test_vapp_template.ova is expected to be present, document it somewhere. It can save some debugging time in future for a new user.
</blockquote></details>

This template is checked into github.

---

*[examples/tenant-onboard.py, line 48 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBWkp_sBHFz9u-kEWOq:-LBWkp_sBHFz9u-kEWOr:b-58fhof) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/examples/tenant-onboard.py#L48)):*
> ```Python
>             pvdc_name = pvdc_ref.get('name')
>             print("Defaulting to first pvdc: {0}".format(pvdc_name))
>             vdc_kwargs['provider_vdc_name'] = pvdc_name
> ```

We should break after processing the first pvdc, and if no pvdcs are found then we should raise an exception.

---

*[examples/tenant-onboard.py, line 57 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBWlJCE3sI6PTQ8xMNi:-LBWlJCE3sI6PTQ8xMNj:b9ffoud) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/examples/tenant-onboard.py#L57)):*
> ```Python
>         netpools = system.list_network_pools()
>         for netpool in netpools:
>             netpool_name = netpool.get('name')
> ```

We should break after processing the first netpool, and if no netpools are found then we should raise an exception.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/234)*
<!-- Sent from Reviewable.io -->

hodgesrm commented:
Review status: 2 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[.Jenkinsfile, line 7 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOHveBDVAmsLrhm1n7:-LBcdAtD4V2IqUR8u7Lu:bnzjzji) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/.Jenkinsfile#L7)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

I believe the email sending code isn't working as of now. Correct me if I am wrong @hodgesr
</blockquote></details>

Removed.  We'll add later on in a separate task. 

---

*[.Jenkinsfile, line 34 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOJRED6WL6sikzlHrW:-LBgquqm627ySmVHNO96:b-qp40zm) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/.Jenkinsfile#L34)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

What is checkout scm doing? git clone? If yes, where is this command or script located?
</blockquote></details>

These are native commands in Jenkins.  I have added a comment as the syntax is really confusing (maybe not to groovy experts but that does not include most of us.)

---

*[.Jenkinsfile, line 96 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOK75a1SvF_RCC1cHN:-LBgqSmMAgMi57V0BAOF:b517ede) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/.Jenkinsfile#L96)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Regardless of success or failure, restoration of testbed must happen.
</blockquote></details>

Fixed.  System test org is cleaned up after a successful test. For failing tests we don't clean up as it allows diagnosis of what went wrong.  We may add more cleanup over time.  

---

*[examples/tenant.yaml, line 55 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOKxcgCtv8cIZu0UtW:-LBcXH--3f0f4l8BqRiX:bf3l4q8) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/examples/tenant.yaml#L55)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

This template is checked into github.
</blockquote></details>

Thanks! Added comment to .yaml and also noted in README.md that scripts must run from this directory so that paths line up. 

---

*[examples/tenant-onboard.py, line 48 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBWkp_sBHFz9u-kEWOq:-LBgqoq88u5EUTAee7Mf:b-lx4fj3) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/examples/tenant-onboard.py#L48)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

We should break after processing the first pvdc, and if no pvdcs are found then we should raise an exception.
</blockquote></details>

Fixed. 

---

*[examples/tenant-onboard.py, line 57 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBWlJCE3sI6PTQ8xMNi:-LBgqqqo63l5XVTX7K4x:bt2vr1b) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/examples/tenant-onboard.py#L57)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

We should break after processing the first netpool, and if no netpools are found then we should raise an exception.
</blockquote></details>

Fixed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/234)*
<!-- Sent from Reviewable.io -->

sahithi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
Robert, I made a comment on clean-up mechanism, but you can re-consider it at later phases of CI/CD.

---

Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[.Jenkinsfile, line 96 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOK75a1SvF_RCC1cHN:-LBvq70rD77CZ6w_rQ99:bnnas1z) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/.Jenkinsfile#L96)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

Fixed.  System test org is cleaned up after a successful test. For failing tests we don't clean up as it allows diagnosis of what went wrong.  We may add more cleanup over time.  
</blockquote></details>

In my opinion, force deleting the org after tests works for now, but it is not an ideal way to restore the test-bed state. Our end goal of clean-up task should be to restore the test-bed to clean state/snapshot (or) create new.

Re-using the same test-bed for really long time may help  catching vCD bugs but will not help catching Pyvcloud bugs. Below are two main reasons why 

1. Anytime there is an issue, we will have to spend more time debugging to identify if this is vCD environmental issue (or) the real Pyvcloud issue. For example, there were several vCD issues in the past on force deletion of orgs/ovdcs/vapps/networks leaving stale entries in vCD DB triggering other unknown issues. we should be mindful of this.

2. Given we are planning to run tests against multiple vCD versions (to ensure Pyvcloud compatibility with older version of vCD), we have all the more reason to start with clean test-beds to get the correct results.

For the same reason, our vCD CTOTS restore the testbed to snapshot for each run. 

---


*Comments from [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/234#-:-LBvuBPW1iNb4gguOsx2:b-njokzu)*
<!-- Sent from Reviewable.io -->

hodgesrm commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[.Jenkinsfile, line 96 at r1](https://reviewable.io/reviews/vmware/pyvcloud/234#-LBOK75a1SvF_RCC1cHN:-LBwNugYAMz3tudkLLFG:b9hu9on) ([raw file](https://github.com/vmware/pyvcloud/blob/4e3933773978b30d4f24a9ed4a4a45e2f2fe6e47/.Jenkinsfile#L96)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

In my opinion, force deleting the org after tests works for now, but it is not an ideal way to restore the test-bed state. Our end goal of clean-up task should be to restore the test-bed to clean state/snapshot (or) create new.

Re-using the same test-bed for really long time may help  catching vCD bugs but will not help catching Pyvcloud bugs. Below are two main reasons why 

1. Anytime there is an issue, we will have to spend more time debugging to identify if this is vCD environmental issue (or) the real Pyvcloud issue. For example, there were several vCD issues in the past on force deletion of orgs/ovdcs/vapps/networks leaving stale entries in vCD DB triggering other unknown issues. we should be mindful of this.

2. Given we are planning to run tests against multiple vCD versions (to ensure Pyvcloud compatibility with older version of vCD), we have all the more reason to start with clean test-beds to get the correct results.

For the same reason, our vCD CTOTS restore the testbed to snapshot for each run. 


</blockquote></details>

I agree--we'll work in getting this to work incrementally.  It gets more interesting as the tests become more complex.  Thanks for the (as always) careful review!

---


*Comments from [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/234)*
<!-- Sent from Reviewable.io -->

rocknes commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 8 files at r1, 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[examples/tenant.yaml, line 52 at r2](https://reviewable.io/reviews/vmware/pyvcloud/234#-LC2tnGRBbWzXl0fhv8x:-LC2tnGRBbWzXl0fhv8y:b-61tk5p) ([raw file](https://github.com/vmware/pyvcloud/blob/b7bbd18ae6a8b18243b57f2fa5308105cf34b053/examples/tenant.yaml#L52)):*
> ```YAML
> # path of files. 
> catalog_items: 
>   - item_name: "Test-Vapp"
> ```

nit : Test-VApp-Template

---


*Comments from [Reviewable](https://reviewable.io/reviews/vmware/pyvcloud/234#-:-LC2vI-3Dr_Pe6Q97YOx:bnfp4nl)*
<!-- Sent from Reviewable.io -->
